xclim/sdba/measures.py

```diff
@@ -0,0 +1,323 @@
+"""
+Measures submodule
+==================
+SDBA diagnostic tests are made up of properties and measures. Measures compare adjusted simulations to a reference,
+through statistical properties or directly.
+This framework for the diagnostic tests was inspired by the [VALUE]_ project.
+
+ .. [VALUE] http://www.value-cost.eu/
+"""
+from typing import Callable
+from warnings import warn
+
+import numpy as np
+import xarray as xr
+from boltons.funcutils import wraps
+from sklearn import metrics
+
+from xclim import sdba
+from xclim.core.formatting import update_xclim_history
+from xclim.core.units import convert_units_to, units2pint
+
+
+def check_same_units_and_convert(func) -> Callable:
+    """Verify that the simulation and the reference have the same units.
+    If not, it converts the simulation to the units of the reference"""
+
+    @wraps(
+        func
+    )  # in order to keep the docstring of the function where the decorator is applied
+    def _check_same_units(*args):
+        sim = args[0]
+        ref = args[1]
+        units_sim = units2pint(sim.units)
+        units_ref = units2pint(ref.units)
+
+        if units_sim != units_ref:
+            warn(
+                f" sim({units_sim}) and ref({units_ref}) don't have the same units."
+                f" sim will be converted to {units_ref}."
+            )
+            sim = convert_units_to(sim, ref)
+        out = func(sim, ref, *args[2:])
+        return out
+
+    return _check_same_units
+
+
+@check_same_units_and_convert
+@update_xclim_history
+def bias(sim: xr.DataArray, ref: xr.DataArray) -> xr.DataArray:
+    """Bias.
+
+    The bias is the simulation minus the reference.
+
+    Parameters
+    ----------
+    sim : xr.DataArray
+      data from the simulation (one value for each grid-point)
+    ref : xr.DataArray
+      data from the reference (observations) (one value for each grid-point)
+
+    Returns
+    -------
+    xr.DataArray,
+      Bias between the simulation and the reference
+
+    Examples
+    --------
+    >>> from xclim.testing import open_dataset
+    >>> tasmax = open_dataset(path_to_tasmax_file).tasmax.isel(time=0)
+    >>> tasmin = open_dataset(path_to_tasmin_file).tasmin.isel(time=0)
+    >>> bias(tasmax, tasmin)
```

Comment:
These examples with tasmin/tasmax don't really make sense. I think that could be a bit confusing. We also don't need an example for every function, even more so when their use is quite straightforward.

Reply:
Ok, I thought Examples were mandatory, so I used the test variables available. I agree that it doesn't make a lot of sense. I will remove the examples.

Reply:
Just as an FYI for future examples, the doctests have a special namespace separate from the module they're written in, with the following imported: `import xclim.testing as xr`. As such, all the doctests use `xr.open_dataset()` so that we can both show users how to typically use the functions, and sneakily run the testing utilities over them.

Reply:
ahh that's neat!